File: apps/ios/Sources/TePlannerKit/Services/FeatureFlags.swift
import Foundation

/// Lightweight feature-flag registry. Each flag has a hard-coded
/// production default and a UserDefaults override key — TestFlight /
/// dev builds can flip flags via the Settings sheet without shipping
/// a new App Store binary.
///
/// Why not a fancier system: we have one flag and 6-week ICP/软著
/// process ahead. A `bool { get }` over UserDefaults is fine.
@MainActor
public enum FeatureFlags {

    public enum Flag: String, CaseIterable {
        /// Hides the Hub "充电规划" entry + the in-flight trip card
        /// while we polish the multi-stop trip pipeline. Two production
        /// bugs surfaced in the same evening (active_trip dead feature,
        /// destination key mismatch); we want to ship a 1.0 without
        /// this feature and re-enable once it's stable. Default OFF.
        case chargingPlanning = "feature.planning.enabled"

        /// Hard-coded production default. iOS reads this when the
        /// UserDefaults key is absent.
        public var defaultValue: Bool {
            switch self {
            case .chargingPlanning: return false
            }
        }

        /// User-facing label for the Settings toggle.
        public var displayName: String {
            switch self {
            case .chargingPlanning: return "充电规划"
            }
        }

        /// One-line explanation under the toggle. Tells testers why
        /// this is gated and what to expect when they flip it.
        public var description: String {
            switch self {
            case .chargingPlanning:
                return "多段充电路线规划与发送至车机。正在打磨中，关闭后 Hub 不显示「充电规划」入口及进行中行程卡片。"
            }
        }
    }

    /// Read a flag. Looks up the UserDefaults override first (so
    /// testers can flip via Settings); falls back to the compiled
    /// default. The key namespace `feature.*` is reserved for this
    /// registry and must never collide with other prefs.
    public static func isOn(_ flag: Flag) -> Bool {
        if let v = UserDefaults.standard.object(forKey: flag.rawValue) as? Bool {
            return v
        }
        return flag.defaultValue
    }

    /// Set the override. Persists across launches. Pass `nil` to
    /// remove the override and fall back to the default — useful for
    /// "reset to defaults" debug actions.
    public static func setOverride(_ flag: Flag, to value: Bool?) {
        if let value {
            UserDefaults.standard.set(value, forKey: flag.rawValue)
        } else {
            UserDefaults.standard.removeObject(forKey: flag.rawValue)
        }
    }

    /// True when this binary is running in TestFlight (sandbox
    /// receipt) or the simulator/dev install (no receipt at all).
    /// Used to gate the Settings toggle so App Store users don't see
    /// debug controls. A separate property — not a flag — because it
    /// describes the *binary*, not user preference.
    public static var isInternalBuild: Bool {
        #if DEBUG
        return true
        #else
        guard let url = Bundle.main.appStoreReceiptURL else {
            // No receipt at all → dev install (e.g. side-loaded). Treat
            // as internal so engineers can still flip flags from the UI.
            return true
        }
        return url.lastPathComponent == "sandboxReceipt"
        #endif
    }
}

File: apps/android/app/src/main/java/cloud/teplanner/android/util/FeatureFlags.kt
package cloud.teplanner.android.util

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit

/**
 * Lightweight feature-flag registry. Each flag has a hard-coded
 * production default and a SharedPreferences override key — internal
 * builds can flip flags via the Settings screen without shipping a
 * new Play Store binary.
 *
 * Mirrors the iOS [FeatureFlags] registry — keep flag rawValues + the
 * default-value table identical so a tester gets the same product
 * shape on both platforms.
 */
object FeatureFlags {

    enum class Flag(val key: String, val default: Boolean, val displayName: String, val description: String) {
        /**
         * Hides the Hub "充电规划" entry while we polish the multi-stop
         * trip pipeline. Two production bugs surfaced in one evening on
         * iOS (dead feature + destination-key mismatch); v1.0 ships
         * without this surface and we re-enable once stable. Default OFF.
         */
        ChargingPlanning(
            key = "feature.planning.enabled",
            default = false,
            displayName = "充电规划",
            description = "多段充电路线规划与发送至车机。正在打磨中，关闭后 Hub 不显示「充电规划」入口。",
        );

        companion object {
            fun all(): List<Flag> = values().toList()
        }
    }

    private const val PREFS_NAME = "teplanner.feature_flags"

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /**
     * Read a flag from a SharedPreferences directly — useful for tests
     * that can't easily construct a Context. Production callers use
     * the [Context] overload.
     */
    fun isOn(prefs: SharedPreferences, flag: Flag): Boolean =
        if (prefs.contains(flag.key)) prefs.getBoolean(flag.key, flag.default) else flag.default

    /**
     * Read a flag. Looks up the SharedPreferences override first (so
     * internal testers can flip via Settings); falls back to the
     * compiled default.
     */
    fun isOn(context: Context, flag: Flag): Boolean = isOn(prefs(context), flag)

    /**
     * Test-friendly variant of [setOverride] that takes the
     * SharedPreferences directly.
     */
    fun setOverride(prefs: SharedPreferences, flag: Flag, value: Boolean?) {
        prefs.edit {
            if (value == null) remove(flag.key) else putBoolean(flag.key, value)
        }
    }

    /**
     * Set the override. Persists across launches. Pass `null` to
     * remove the override and fall back to the default.
     */
    fun setOverride(context: Context, flag: Flag, value: Boolean?) =
        setOverride(prefs(context), flag, value)

    /**
     * True for debug builds and internal-track installs. Matches iOS
     * `FeatureFlags.isInternalBuild` semantics — gates whether the
     * Settings screen shows the flag-override section.
     */
    fun isInternalBuild(): Boolean = BuildConfigShim.DEBUG

    /**
     * Indirection so we don't hard-import the app-module's
     * `BuildConfig` from this util (kept in :app, callers pass the
     * boolean through if they prefer). DEBUG defaults to true; release
     * builds override via [setInternalBuildFlag].
     */
    private object BuildConfigShim {
        @Volatile
        var DEBUG: Boolean = true
    }

    /** Called from Application.onCreate() with [BuildConfig.DEBUG]. */
    fun setInternalBuildFlag(debug: Boolean) {
        BuildConfigShim.DEBUG = debug
    }
}
